src/Microsoft.ML.PCA/PcaTransform.cs

```diff
@@ -493,56 +530,134 @@ private Float[][] PostProcess(Float[][] y, Float[] sigma, Float[] z, int d, int
+                        }
+                        return index;
+                    }
+                }
+            }
 
-            var values = dst.Values;
-            if (Utils.Size(values) < transformInfo.Rank)
-                values = new Float[transformInfo.Rank];
+            private readonly PcaTransform _parent;
+            private readonly int _numColumns;
 
-            for (int i = 0; i < transformInfo.Rank; i++)
+            public Mapper(PcaTransform parent, Schema inputSchema)
+               : base(parent.Host.Register(nameof(Mapper)), parent, inputSchema)
             {
-                values[i] = VectorUtils.DotProductWithOffset(transformInfo.Eigenvectors[i], 0, ref src) -
-                    (transformInfo.MeanProjected == null ? 0 : transformInfo.MeanProjected[i]);
+                _parent = parent;
+                _numColumns = parent._numColumns;
+                for (int i = 0; i < _numColumns; i++)
+                {
+                    var colPair = _parent.ColumnPairs[i];
+                    var colSchemaInfo = new ColumnSchemaInfo(colPair, inputSchema);
+                    ValidatePcaInput(Host, colPair.input, colSchemaInfo.InputType);
+                    if (colSchemaInfo.InputType.VectorSize != _parent._transformInfos[i].Dimension)
+                    {
+                        var msg = $"Dimension of column ${colPair.input} is ${colSchemaInfo.InputType.VectorSize}, which doesn't match the expected size ${_parent._transformInfos[i].Dimension}";
+                        throw Host.Except(msg);
+                    }
+                }
             }
 
-            dst = new VBuffer<Float>(transformInfo.Rank, values, dst.Indices);
+            public override Schema.Column[] GetOutputColumns()
+            {
+                var result = new Schema.Column[_numColumns];
+                for (int i = 0; i < _numColumns; i++)
+                    result[i] = new Schema.Column(_parent.ColumnPairs[i].output, _parent._transformInfos[i].OutputType, null);
+                return result;
+            }
+
+            protected override Delegate MakeGetter(IRow input, int iinfo, out Action disposer)
+            {
+                Contracts.AssertValue(input);
+                Contracts.Assert(0 <= iinfo && iinfo < _numColumns);
+                disposer = null;
+
+                var srcGetter = input.GetGetter<VBuffer<float>>(ColMapNewToOld[iinfo]);
+                var src = default(VBuffer<float>);
+
+                ValueGetter<VBuffer<float>> dstGetter = (ref VBuffer<float> dst) =>
+                    {
+                        srcGetter(ref src);
+                        TransformFeatures(Host, ref src, ref dst, _parent._transformInfos[iinfo]);
+                    };
+
+                return dstGetter;
+            }
+
+            private static void TransformFeatures(IExceptionContext ectx, ref VBuffer<float> src, ref VBuffer<float> dst, TransformInfo transformInfo)
```

Comment:
>static void TransformFeatures [](start = 20, length = 29)

nit: I don't see reason why it's static, you can just make it part of class and stop accept IExceptionContext and use Host instead

Reply:
Or keep as is :)

---
In reply to: [227575467](https://github.com/dotnet/machinelearning/pull/1333#discussion_r227575467) [](ancestors = 227575467)